refactor: Add EndpointContext

lqiu96 opened the pull request:
Thank you for opening a Pull Request! For general contributing guidelines, please refer to [contributing guide](https://github.com/googleapis/gapic-generator-java/blob/main/CONTRIBUTING.md)

Before submitting your PR, there are a few things you can do to make sure it goes smoothly:

- [x] Make sure to open an issue as a [bug/issue](https://github.com/googleapis/gapic-generator-java/issues/new/choose) before writing your code! That way we can discuss the change, evaluate designs, and agree on the general idea
- [x] Ensure the tests and linter pass
- [x] Code coverage does not decrease (if any source code was changed)
- [x] Appropriate docs were updated (if necessary)

## Description
EndpointContext is a new class created to resolve the correct endpoint. Client library users are able customize the endpoint in various ways and this class is the source of truth to resolve all the customizations.

Currently supports:
- ClientSettings/ StubSettings endpoint value
- mTLS endpoint value
- switchToMtls flag

The logic to resolve the endpoint was moved from the ClientContext into the EndpointContext. In this PR, no additional logic was added.

blakeli0 commented:
Can we integrate EndpointContext into the existing code and make sure all functionalities are not affected? It may not be a good practice to have a `chore` PR introduces new classes without any reference to it. Because it does not provide  additional functionality, and it's not easy for reviewers to understand the purpose of the new class. It would be better if we can make this PR a proper `refactor`.

lqiu96 commented on gax-java/gax/src/main/java/com/google/api/gax/rpc/EndpointContext.java:
```diff
@@ -0,0 +1,129 @@
+
+    public abstract Builder setSwitchToMtlsEndpointAllowed(boolean switchToMtlsEndpointAllowed);
+
+    @VisibleForTesting
```
Yep. Removed.

blakeli0 commented on gax-java/gax/src/main/java/com/google/api/gax/rpc/EndpointContext.java:
```diff
@@ -0,0 +1,126 @@
+  public abstract String clientSettingsEndpoint();
+
+  @Nullable
+  public abstract String transportChannelEndpoint();
```
Will we ever need to pass in `transportChannelEndpoint`? 
I see that there is no getters for the endpoint set to transport channel, maybe we want to expose one? So that we can set all the info to `EndpointContext`?
It's also OK to leave `transportChannelEndpoint` out of `EndpointContext` for now, if that's the case, then it's better to remove this field at this moment.

lqiu96 replied:
> Will we ever need to pass in transportChannelEndpoint?

I believe so to keep with the existing behavior where transportChannelEndpoint overrides the clientSettingsEndpoint if set. Right now, the ClientContext doesn't set it inside the EndpointContext but the code to use the endpoint exists here: https://github.com/googleapis/sdk-platform-java/blob/da6607b17130ab045640618d505fda915ddb8e49/gax-java/gax/src/main/java/com/google/api/gax/rpc/ClientContext.java#L233-L235

The reason I haven't added it in yet is because I _most likely_ will need to expose an additional `getEndpoint()` method to the TransportChannelProvider: https://github.com/googleapis/sdk-platform-java/blob/da6607b17130ab045640618d505fda915ddb8e49/gax-java/gax/src/main/java/com/google/api/gax/rpc/TransportChannelProvider.java#L144 (Note: grpc and httpjson's transportchannel already have a `getEndpoint()` method). I was planning on doing this on the next PR as this one only introduces the EndpointContext.

I see your point about including it then, so I'll drop the transportChannelEndpoint values from this PR and add it in the next PR.

blakeli0 commented on gax-java/gax/src/main/java/com/google/api/gax/rpc/EndpointContext.java:
```diff
@@ -0,0 +1,126 @@
+    return endpoint;
+  }
+
+  public String resolveEndpoint() throws IOException {
```
Can we try to resolve the endpoint in the `build()` method and just expose a getter for the resolved endpoint?

lqiu96 replied:
I'm using Autovalue's Builder which I believe generates the `build()` implementation for us. I could create my own Builder class if that's preferable.

blakeli0 replied:
Yeah, it's a common practice to validate or normalize a field value in the `build()` method, see internal docs [here](go/autovalue/builders-howto#normalize). We have [examples](https://github.com/googleapis/sdk-platform-java/blob/main/gax-java/gax-grpc/src/main/java/com/google/api/gax/grpc/ChannelPoolSettings.java#L163-L181) in our repo as well.

blakeli0 commented on gax-java/gax/src/main/java/com/google/api/gax/rpc/EndpointContext.java:
```diff
@@ -0,0 +1,126 @@
+@AutoValue
+public abstract class EndpointContext {
+  @Nullable
+  public abstract String clientSettingsEndpoint();
```
I don't think this can be null as it is coming from [StubSettings.endpoint](https://github.com/googleapis/sdk-platform-java/blob/main/gax-java/gax/src/main/java/com/google/api/gax/rpc/StubSettings.java#L73)?

lqiu96 replied:
I believe for most non-modified GAPICs, this is the case. It should be set via the {Service}StubSettings like: https://github.com/googleapis/google-cloud-java/blob/6227504db7c511b379cad2d67960d63f36eed39b/java-grafeas/src/main/java/io/grafeas/v1/stub/GrafeasStubSettings.java#L635 with a non-null value (as the generator will fail if it can't find it).

I'm accounting for the Grafeas edge case: https://github.com/googleapis/google-cloud-java/blob/6227504db7c511b379cad2d67960d63f36eed39b/java-grafeas/src/main/java/io/grafeas/v1/stub/GrafeasStubSettings.java#L403-L405 which has post-processed changes.

blakeli0 replied:
I see, this is good for now then. Let's keep this in mind and come back for it later if we have time. I'm not sure why we removed the default endpoint for grafeas, and how does grafeas create channels without a default endpoint. But ideally this should not be nullable, otherwise gRPC channel creation could fail.

blakeli0 commented on gax-java/gax/src/main/java/com/google/api/gax/rpc/EndpointContext.java:
```diff
@@ -0,0 +1,126 @@
+  void determineEndpoint() throws IOException {
+    MtlsProvider mtlsProvider = mtlsProvider() == null ? new MtlsProvider() : mtlsProvider();
+    String customEndpoint =
+        transportChannelEndpoint() != null ? transportChannelEndpoint() : clientSettingsEndpoint();
```
I don't see `transportChannelEndpoint` being set in the ClientContext. Was it intentional?

lqiu96 replied:
I believe this should be answered above. I'll remove this logic from this PR and add it in the next.

blakeli0 replied:
SGTM

blakeli0 commented on gax-java/gax/src/main/java/com/google/api/gax/rpc/ClientContext.java:
```diff
@@ -224,12 +200,13 @@ public static ClientContext create(StubSettings settings) throws IOException {
+            .setMtlsEndpoint(settings.getMtlsEndpoint())
+            .setSwitchToMtlsEndpointAllowed(settings.getSwitchToMtlsEndpointAllowed())
+            .build();
+    String endpoint = endpointContext.resolveEndpoint();
```
It maybe outside the scope of this PR, do we plan to pass the whole `EndpointContext` to `ApiCallContext` so that we can do the validation later?

lqiu96 replied:
Yes, it'll be in a future PR.

lqiu96 commented on gax-java/gax/src/main/java/com/google/api/gax/rpc/EndpointContext.java:
```diff
@@ -0,0 +1,122 @@
+
+    abstract EndpointContext autoBuild();
+
+    public EndpointContext build() {
+      try {
+        EndpointContext endpointContext = autoBuild();
+        endpointContext.determineEndpoint();
+        return endpointContext;
+      } catch (IOException e) {
+        throw new RuntimeException("Unable to determine the endpoint: " + e.getMessage());
```
throw IOExcpetion

lqiu96 commented on gax-java/gax/src/main/java/com/google/api/gax/rpc/EndpointContext.java:
```diff
@@ -0,0 +1,122 @@
+
+  @AutoValue.Builder
+  public abstract static class Builder {
+    public abstract Builder setClientSettingsEndpoint(String clientSettingsEndpoint);
```
Add Javadoc for this field (in builder and setter)

lqiu96 commented on gax-java/gax/src/main/java/com/google/api/gax/rpc/EndpointContext.java:
```diff
@@ -0,0 +1,122 @@
+    return endpoint;
+  }
+
+  public String getResolvedEndpoint() {
+    return resolvedEndpoint;
+  }
```
Add Javadoc here

blakeli0 reviewed (approved):
LGTM. Before merging, can you please add more info to the description to explain the purpose of `EndpointContext`? 
I would reword the title as well, I think this is more like a `refactor` instead of `feat`. Because we added a new class to encapsulate existing features but didn't introduce any new features.

lqiu96 commented:
> LGTM. Before merging, can you please add more info to the description to explain the purpose of `EndpointContext`? I would reword the title as well, I think this is more like a `refactor` instead of `feat`. Because we added a new class to encapsulate existing features but didn't introduce any new features.

Added.
